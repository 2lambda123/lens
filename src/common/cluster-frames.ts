/**
 * Copyright (c) 2021 OpenLens Authors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { action, observable } from "mobx";
import type { ClusterId } from "./cluster-store";
import { iter, Singleton } from "./utils";

export interface ClusterFrameInfo {
  frameId: number;
  processId: number;
  windowId: number;
}

export class ClusterFrames extends Singleton {
  private mapping = observable.map<ClusterId, ClusterFrameInfo>();

  public getAllFrameInfo(): ClusterFrameInfo[] {
    return [...this.mapping.values()];
  }

  public set(clusterId: ClusterId, info: ClusterFrameInfo): void {
    this.mapping.set(clusterId, info);
  }

  public getFrameInfoByClusterId(clusterId: ClusterId): ClusterFrameInfo | undefined {
    return this.mapping.get(clusterId);
  }

  public getFrameInfoByFrameId(frameId: number): ClusterFrameInfo | undefined {
    return iter.find(this.mapping.values(), frameInfo => frameInfo.frameId === frameId);
  }

  public clearInfoForCluster(clusterId: ClusterId): void {
    this.mapping.delete(clusterId);
  }

  @action
  public clearInfoForWindow(windowId: number): void {
    for (const [clusterId, frameInfo] of this.mapping) {
      if (frameInfo.windowId === windowId) {
        this.mapping.delete(clusterId);
      }
    }
  }
}
