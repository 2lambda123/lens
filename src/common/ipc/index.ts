export * from "./ipc";
export * from "./update-available";
export * from "./cluster";
export * from "./type-enforced-ipc";
