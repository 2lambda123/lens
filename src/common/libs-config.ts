// Global configuration setup for external packages.
// Should be imported at the top of app's entry points.
import { configure } from "mobx";
import { enableMapSet, setAutoFreeze } from "immer";

// Mobx
configure({
  isolateGlobalState: true, // allow to use different versions of mobx in app & extensions
  enforceActions: "never", // skip usage of @action for class methods
});

// Immer
setAutoFreeze(false); // allow to merge observables
enableMapSet(); // allow to merge Map() and Set()
