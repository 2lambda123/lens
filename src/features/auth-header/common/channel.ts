/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import { getRequestChannel } from "../../../common/utils/channel/get-request-channel";

export const authHeaderChannel = getRequestChannel<void, string>("auth-header-value");
