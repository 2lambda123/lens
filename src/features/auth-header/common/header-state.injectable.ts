/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import { getInjectable } from "@ogre-tools/injectable";

const authHeaderStateInjectable = getInjectable({
  id: "auth-header-state",
  instantiate: () => {
    let header: string;

    return {
      get: () => {
        if (!header) {
          throw new Error("Tried to get auth header value before being set");
        }

        return header;
      },
      set: (value: string) => {
        if (header) {
          throw new Error("Tried to set auth header value more than once");
        }

        header = value;
      },
    };
  },
});

export default authHeaderStateInjectable;
