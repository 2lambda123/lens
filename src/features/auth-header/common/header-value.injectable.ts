/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import { getInjectable } from "@ogre-tools/injectable";
import authHeaderStateInjectable from "./header-state.injectable";

const authHeaderValueInjectable = getInjectable({
  id: "auth-header-value",
  instantiate: (di) => `Bearer ${di.inject(authHeaderStateInjectable).get()}`,
});

export default authHeaderValueInjectable;
