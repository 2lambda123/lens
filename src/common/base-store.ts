import path from "path";
import Config from "conf";
import { Options as ConfOptions } from "conf/dist/source/types";
import { app, ipcMain, IpcMainEvent, ipcRenderer, IpcRendererEvent, remote } from "electron";
import { IReactionOptions, observable, reaction, when } from "mobx";
import Singleton from "./utils/singleton";
import { getAppVersion } from "./utils/app-version";
import logger from "../main/logger";
import { createTypedSender } from "./ipc";
import isEqual from "lodash/isEqual";
import { autobind } from "./utils";

export interface BaseStoreParams<T = any> extends ConfOptions<T> {
  autoLoad?: boolean;
  syncEnabled?: boolean;
  syncOptions?: IReactionOptions;
}

/**
 * Note: T should only contain base JSON serializable types.
 */
export abstract class BaseStore<T = any> extends Singleton {
  protected storeConfig: Config<T>;
  protected syncDisposers: Function[] = [];

  whenLoaded = when(() => this.isLoaded);
  @observable isLoaded = false;

  protected constructor(protected params: BaseStoreParams) {
    super();
    this.params = {
      autoLoad: false,
      syncEnabled: true,
      ...params,
    };
    this.init();
  }

  get name() {
    return path.basename(this.storeConfig.path);
  }

  protected readonly syncRenderer = createTypedSender({
    channel: `store-sync-renderer:${this.path}`,
    verifier: (src: unknown): src is any => true,
  });

  protected readonly syncMain = createTypedSender({
    channel: `store-sync-main:${this.path}`,
    verifier: (src: unknown): src is any => true,
  });

  get path() {
    return this.storeConfig.path;
  }

  protected async init() {
    if (this.params.autoLoad) {
      await this.load();
    }

    if (this.params.syncEnabled) {
      await this.whenLoaded;
      this.enableSync();
    }
  }

  async load() {
    const { autoLoad, syncEnabled, ...confOptions } = this.params;

    this.storeConfig = new Config({
      ...confOptions,
      projectName: "lens",
      projectVersion: getAppVersion(),
      cwd: this.cwd(),
    });
    logger.info(`[STORE]: LOADED from ${this.path}`);
    this.fromStore(this.storeConfig.store);
    this.isLoaded = true;
  }

  protected cwd() {
    return (app || remote.app).getPath("userData");
  }

  protected async saveToFile(model: T) {
    logger.info(`[STORE]: SAVING ${this.path}`);
    // todo: update when fixed https://github.com/sindresorhus/conf/issues/114
    Object.entries(model).forEach(([key, value]) => {
      this.storeConfig.set(key, value);
    });
  }

  enableSync() {
    this.syncDisposers.push(
      reaction(() => this.toJSON(), this.onModelChange, this.params.syncOptions),
    );

    if (ipcMain) {
      this.syncDisposers.push(this.syncMain.on((event: IpcMainEvent, model: T) => {
        logger.silly(`[STORE]: SYNC ${this.name} from renderer`, { model });
        this.onSync(model);
      }));
    }

    if (ipcRenderer) {
      this.syncDisposers.push(this.syncRenderer.on((event: IpcRendererEvent, model: T) => {
        logger.silly(`[STORE]: SYNC ${this.name} from main`, { model });
        this.disableSync();
        this.onSync(model);

        if (this.params.syncEnabled) {
          this.enableSync();
        }
      }));
    }
  }

  disableSync() {
    this.syncDisposers.forEach(dispose => dispose());
    this.syncDisposers.length = 0;
  }

  protected onSync(model: T) {
    // todo: use "resourceVersion" if merge required (to avoid equality checks => better performance)
    if (!isEqual(this.toJSON(), model)) {
      this.fromStore(model);
    }
  }

  @autobind()
  protected async onModelChange(model: T) {
    if (ipcMain) {
      this.saveToFile(model); // save config file
      this.syncRenderer.broadcast(model);
    } else {
      this.syncMain.broadcast(model);
    }
  }

  /**
   * fromStore is called internally when a child class syncs with the file
   * system.
   * @param data the parsed information read from the stored JSON file
   */
  protected abstract fromStore(data: T): void;

  /**
   * toJSON is called when syncing the store to the filesystem. It should
   * produce a JSON serializable object representaion of the current state.
   *
   * It is recommended that a round trip is valid. Namely, calling
   * `this.fromStore(this.toJSON())` shouldn't change the state.
   */
  abstract toJSON(): T;
}
