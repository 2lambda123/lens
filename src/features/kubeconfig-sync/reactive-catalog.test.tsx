/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import type { DiContainer } from "@ogre-tools/injectable";
import type { RenderResult } from "@testing-library/react";
import navigateToCatalogInjectable from "../../common/front-end-routing/routes/catalog/navigate-to-catalog.injectable";
import writeFileInjectable from "../../common/fs/write-file.injectable";
import { dumpConfigYaml } from "../../common/kube-helpers";
import homeDirectoryPathInjectable from "../../common/os/home-directory-path.injectable";
import joinPathsInjectable from "../../common/path/join-paths.injectable";
import { flushPromises } from "../../common/test-utils/flush-promises";
import type { ApplicationBuilder } from "../../renderer/components/test-utils/get-application-builder";
import { getApplicationBuilder } from "../../renderer/components/test-utils/get-application-builder";

describe("kubeconfig sync showing reactive catalog", () => {
  let builder: ApplicationBuilder;
  let rendered: RenderResult;
  let windowDi: DiContainer;
  let mainDi: DiContainer;

  beforeEach(async () => {
    builder = getApplicationBuilder();

    // builder.mainDi.override(loggerInjectable, () => console as any);
    rendered = await builder.render();
    windowDi = builder.applicationWindow.only.di;
    mainDi = builder.mainDi;
  });

  it("renders", () => {
    expect(rendered.baseElement).toMatchSnapshot();
  });

  describe("when navigating to the catalog", () => {
    beforeEach(() => {
      const navigateToCatalog = windowDi.inject(navigateToCatalogInjectable);

      navigateToCatalog();
    });

    it("renders", () => {
      expect(rendered.baseElement).toMatchSnapshot();
    });

    describe("when a config file is written under ~/.kube", () => {
      beforeEach(async () => {
        const writeFile = mainDi.inject(writeFileInjectable);
        const joinPaths = mainDi.inject(joinPathsInjectable);
        const homeDirectoryPath = mainDi.inject(homeDirectoryPathInjectable);

        const configContents = dumpConfigYaml({
          clusters: [{
            name: "some-cluster-name",
            server: "https://1.2.3.4",
            skipTLSVerify: false,
          }],
          users: [{
            name: "some-user-name",
          }],
          contexts: [{
            cluster: "some-cluster-name",
            name: "some-context-name",
            user: "some-user-name",
          }],
        });

        await writeFile(joinPaths(homeDirectoryPath, ".kube", "config"), configContents);
        await flushPromises();
      });

      it.only("eventually shows the cluster as a new entity", async () => {
        await rendered.findByTestId("catalog-entity-row-for-some-cluster-name", undefined, {
          timeout: 10_000,
        });
      }, 100_000);
    });
  });
});
