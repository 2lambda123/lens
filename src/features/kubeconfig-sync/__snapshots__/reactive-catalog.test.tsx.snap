// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kubeconfig sync showing reactive catalog renders 1`] = `
<body>
  <div>
    <div
      class="ClusterManager"
    >
      <div
        class="topBar"
      >
        <div
          class="items"
        >
          <div
            class="preventedDragging"
          >
            <i
              class="Icon material interactive disabled focusable"
              data-testid="home-button"
            >
              <span
                class="icon"
                data-icon-name="home"
              >
                home
              </span>
            </i>
          </div>
          <div
            class="size-sm"
          />
          <div
            class="preventedDragging"
          >
            <i
              class="Icon material interactive disabled focusable"
              data-testid="history-back"
            >
              <span
                class="icon"
                data-icon-name="arrow_back"
              >
                arrow_back
              </span>
            </i>
          </div>
          <div
            class="size-sm"
          />
          <div
            class="preventedDragging"
          >
            <i
              class="Icon material interactive disabled focusable"
              data-testid="history-forward"
            >
              <span
                class="icon"
                data-icon-name="arrow_forward"
              >
                arrow_forward
              </span>
            </i>
          </div>
          <div
            class="separator"
          />
        </div>
      </div>
      <main>
        <div
          id="lens-views"
        />
        <div
          class="flex justify-center Welcome align-center"
          data-testid="welcome-page"
        >
          <div
            data-testid="welcome-banner-container"
            style="width: 320px;"
          >
            <i
              class="Icon logo svg focusable"
            >
              <span
                class="icon"
              />
            </i>
            <div
              class="flex justify-center"
            >
              <div
                data-testid="welcome-text-container"
                style="width: 320px;"
              >
                <h2>
                  Welcome to some-product-name!
                </h2>
                <p>
                  To get you started we have auto-detected your clusters in your
                   
                  kubeconfig file and added them to the catalog, your centralized
                   
                  view for managing all your cloud-native resources.
                  <br />
                  <br />
                  If you have any questions or feedback, please join our 
                  <a
                    class="link"
                    href="https://k8slens.dev/slack.html"
                    rel="noreferrer"
                    target="_blank"
                  >
                    Lens Community slack channel
                  </a>
                  .
                </p>
                <ul
                  class="block"
                  data-testid="welcome-menu-container"
                  style="width: 320px;"
                >
                  <li
                    class="flex grid-12"
                  >
                    <i
                      class="Icon box col-1 material focusable"
                    >
                      <span
                        class="icon"
                        data-icon-name="view_list"
                      >
                        view_list
                      </span>
                    </i>
                    <a
                      class="box col-10"
                    >
                      Browse Clusters in Catalog
                    </a>
                    <i
                      class="Icon box col-1 material focusable"
                    >
                      <span
                        class="icon"
                        data-icon-name="navigate_next"
                      >
                        navigate_next
                      </span>
                    </i>
                  </li>
                </ul>
              </div>
            </div>
          </div>
        </div>
      </main>
      <div
        class="HotbarMenu flex column"
      >
        <div
          class="HotbarItems flex column gaps"
        >
          <div
            class="HotbarCell isDraggingOwner animateDown"
            index="0"
          >
            <div
              style="z-index: 12; position: absolute;"
            >
              <div
                class="HotbarIcon contextMenuAvailable"
              >
                <div
                  class="Avatar rounded disabled avatar"
                  id="hotbarIcon-hotbar-icon-catalog-entity"
                  style="width: 40px; height: 40px; background: rgb(5, 1, 130);"
                >
                  Ca
                </div>
              </div>
            </div>
          </div>
          <div
            class="HotbarCell isDraggingOwner animateDown"
            index="1"
          />
          <div
            class="HotbarCell isDraggingOwner animateDown"
            index="2"
          />
          <div
            class="HotbarCell isDraggingOwner animateDown"
            index="3"
          />
          <div
            class="HotbarCell isDraggingOwner animateDown"
            index="4"
          />
          <div
            class="HotbarCell isDraggingOwner animateDown"
            index="5"
          />
          <div
            class="HotbarCell isDraggingOwner animateDown"
            index="6"
          />
          <div
            class="HotbarCell isDraggingOwner animateDown"
            index="7"
          />
          <div
            class="HotbarCell isDraggingOwner animateDown"
            index="8"
          />
          <div
            class="HotbarCell isDraggingOwner animateDown"
            index="9"
          />
          <div
            class="HotbarCell isDraggingOwner animateDown"
            index="10"
          />
          <div
            class="HotbarCell isDraggingOwner animateDown"
            index="11"
          />
        </div>
        <div
          class="HotbarSelector"
        >
          <i
            class="Icon Icon previous material interactive focusable"
            tabindex="0"
          >
            <span
              class="icon"
              data-icon-name="arrow_left"
            >
              arrow_left
            </span>
          </i>
          <div
            class="HotbarIndex"
          >
            <div
              class="badge Badge small clickable"
              id="hotbarIndex"
            >
              1
            </div>
          </div>
          <i
            class="Icon material interactive focusable"
            tabindex="0"
          >
            <span
              class="icon"
              data-icon-name="arrow_right"
            >
              arrow_right
            </span>
          </i>
        </div>
      </div>
      <div
        class="StatusBar"
        data-testid="status-bar"
      >
        <div
          class="leftSide"
          data-testid="status-bar-left"
        />
        <div
          class="rightSide"
          data-testid="status-bar-right"
        />
      </div>
    </div>
    <div
      class="Notifications flex column align-flex-end"
    />
  </div>
</body>
`;

exports[`kubeconfig sync showing reactive catalog when navigating to the catalog renders 1`] = `
<body>
  <div>
    <div
      class="ClusterManager"
    >
      <div
        class="topBar"
      >
        <div
          class="items"
        >
          <div
            class="preventedDragging"
          >
            <i
              class="Icon material interactive focusable"
              data-testid="home-button"
              tabindex="0"
            >
              <span
                class="icon"
                data-icon-name="home"
              >
                home
              </span>
            </i>
          </div>
          <div
            class="size-sm"
          />
          <div
            class="preventedDragging"
          >
            <i
              class="Icon material interactive disabled focusable"
              data-testid="history-back"
            >
              <span
                class="icon"
                data-icon-name="arrow_back"
              >
                arrow_back
              </span>
            </i>
          </div>
          <div
            class="size-sm"
          />
          <div
            class="preventedDragging"
          >
            <i
              class="Icon material interactive disabled focusable"
              data-testid="history-forward"
            >
              <span
                class="icon"
                data-icon-name="arrow_forward"
              >
                arrow_forward
              </span>
            </i>
          </div>
          <div
            class="separator"
          />
        </div>
      </div>
      <main>
        <div
          id="lens-views"
        />
        <div
          class="mainLayout"
          style="--sidebar-width: 200px;"
        >
          <div
            class="sidebar"
          >
            <div
              class="flex flex-col w-full"
            >
              <div
                class="catalog"
              >
                Catalog
              </div>
              <ul
                aria-multiselectable="false"
                class="MuiTreeView-root"
                role="tree"
              >
                <li
                  aria-selected="true"
                  class="MuiTreeItem-root Mui-selected"
                  data-testid="*-tab"
                  role="treeitem"
                  tabindex="0"
                >
                  <div
                    class="MuiTreeItem-content"
                  >
                    <div
                      class="MuiTreeItem-iconContainer"
                    />
                    <div
                      class="MuiTypography-root MuiTreeItem-label MuiTypography-body1"
                    >
                      Browse
                    </div>
                  </div>
                </li>
                <li
                  aria-expanded="true"
                  class="MuiTreeItem-root bordered Mui-expanded"
                  role="treeitem"
                  tabindex="-1"
                >
                  <div
                    class="MuiTreeItem-content"
                  >
                    <div
                      class="MuiTreeItem-iconContainer"
                    >
                      <i
                        class="Icon material focusable"
                      >
                        <span
                          class="icon"
                          data-icon-name="expand_more"
                        >
                          expand_more
                        </span>
                      </i>
                    </div>
                    <div
                      class="MuiTypography-root MuiTreeItem-label MuiTypography-body1"
                    >
                      <div
                        class="parent"
                      >
                        Categories
                      </div>
                    </div>
                  </div>
                  <ul
                    class="MuiCollapse-root MuiTreeItem-group MuiCollapse-entered"
                    role="group"
                    style="min-height: 0px;"
                  >
                    <div
                      class="MuiCollapse-wrapper"
                    >
                      <div
                        class="MuiCollapse-wrapperInner"
                      >
                        <li
                          class="MuiTreeItem-root"
                          data-testid="entity.k8slens.dev/General-tab"
                          role="treeitem"
                          tabindex="-1"
                        >
                          <div
                            class="MuiTreeItem-content"
                          >
                            <div
                              class="MuiTreeItem-iconContainer"
                            >
                              <i
                                class="Icon material focusable small"
                              >
                                <span
                                  class="icon"
                                  data-icon-name="settings"
                                >
                                  settings
                                </span>
                              </i>
                            </div>
                            <div
                              class="MuiTypography-root MuiTreeItem-label MuiTypography-body1"
                            >
                              <div
                                class="flex"
                              >
                                <div>
                                  General
                                </div>
                              </div>
                            </div>
                          </div>
                        </li>
                        <li
                          class="MuiTreeItem-root"
                          data-testid="entity.k8slens.dev/KubernetesCluster-tab"
                          role="treeitem"
                          tabindex="-1"
                        >
                          <div
                            class="MuiTreeItem-content"
                          >
                            <div
                              class="MuiTreeItem-iconContainer"
                            >
                              <i
                                class="Icon focusable small"
                              >
                                <span
                                  class="icon"
                                  data-icon-name=""
                                />
                              </i>
                            </div>
                            <div
                              class="MuiTypography-root MuiTreeItem-label MuiTypography-body1"
                            >
                              <div
                                class="flex"
                              >
                                <div>
                                  Clusters
                                </div>
                              </div>
                            </div>
                          </div>
                        </li>
                        <li
                          class="MuiTreeItem-root"
                          data-testid="entity.k8slens.dev/WebLink-tab"
                          role="treeitem"
                          tabindex="-1"
                        >
                          <div
                            class="MuiTreeItem-content"
                          >
                            <div
                              class="MuiTreeItem-iconContainer"
                            >
                              <i
                                class="Icon material focusable small"
                              >
                                <span
                                  class="icon"
                                  data-icon-name="public"
                                >
                                  public
                                </span>
                              </i>
                            </div>
                            <div
                              class="MuiTypography-root MuiTreeItem-label MuiTypography-body1"
                            >
                              <div
                                class="flex"
                              >
                                <div>
                                  Web Links
                                </div>
                              </div>
                            </div>
                          </div>
                        </li>
                      </div>
                    </div>
                  </ul>
                </li>
              </ul>
            </div>
            <div
              class="ResizingAnchor horizontal trailing"
            />
          </div>
          <div
            class="contents"
          >
            <div
              class="views"
            >
              <div
                class="ItemListLayout flex column Catalog"
                data-testid="catalog-list-for-browse-all"
              >
                <div
                  class="header flex gaps align-center"
                >
                  <h5
                    class="title"
                  >
                    Browse All
                  </h5>
                  <div
                    class="info-panel box grow"
                  >
                    0 items
                  </div>
                  <div
                    class="Input SearchInput focused"
                  >
                    <label
                      class="input-area flex gaps align-center"
                      id=""
                    >
                      <input
                        class="input box grow"
                        placeholder="Search..."
                        spellcheck="false"
                        value=""
                      />
                      <i
                        class="Icon material focusable small"
                      >
                        <span
                          class="icon"
                          data-icon-name="search"
                        >
                          search
                        </span>
                      </i>
                    </label>
                    <div
                      class="input-info flex gaps"
                    />
                  </div>
                </div>
                <div
                  class="items box grow flex column"
                >
                  <div
                    class="Table flex column Catalog box grow dark selectable scrollable sortable autoSize virtual"
                  >
                    <div
                      class="TableHead sticky nowrap topLine"
                    >
                      <div
                        class="TableCell entityName nowrap sorting"
                        id="name"
                      >
                        <div
                          class="content"
                        >
                          Name
                        </div>
                        <i
                          class="Icon sortIcon material focusable"
                        >
                          <span
                            class="icon"
                            data-icon-name="arrow_drop_down"
                          >
                            arrow_drop_down
                          </span>
                        </i>
                      </div>
                      <div
                        class="TableCell nowrap sorting"
                        id="kind"
                      >
                        <div
                          class="content"
                        >
                          Kind
                        </div>
                        <i
                          class="Icon sortIcon material focusable"
                        >
                          <span
                            class="icon"
                            data-icon-name="arrow_drop_down"
                          >
                            arrow_drop_down
                          </span>
                        </i>
                      </div>
                      <div
                        class="TableCell sourceCell nowrap sorting"
                        id="source"
                      >
                        <div
                          class="content"
                        >
                          Source
                        </div>
                        <i
                          class="Icon sortIcon material focusable"
                        >
                          <span
                            class="icon"
                            data-icon-name="arrow_drop_down"
                          >
                            arrow_drop_down
                          </span>
                        </i>
                      </div>
                      <div
                        class="TableCell labelsCell scrollable nowrap"
                        id="labels"
                      >
                        <div
                          class="content"
                        >
                          Labels
                        </div>
                      </div>
                      <div
                        class="TableCell statusCell nowrap sorting"
                        id="status"
                      >
                        <div
                          class="content"
                        >
                          Status
                        </div>
                        <i
                          class="Icon sortIcon material focusable"
                        >
                          <span
                            class="icon"
                            data-icon-name="arrow_drop_down"
                          >
                            arrow_drop_down
                          </span>
                        </i>
                      </div>
                      <div
                        class="TableCell menu nowrap"
                      >
                        <div
                          class="content"
                        >
                          <i
                            class="Icon material interactive focusable"
                            id="menu-actions-for-item-object-list-content"
                            tabindex="0"
                          >
                            <span
                              class="icon"
                              data-icon-name="more_vert"
                            >
                              more_vert
                            </span>
                          </i>
                        </div>
                      </div>
                    </div>
                    <div
                      class="NoItems flex box grow"
                    >
                      <div
                        class="box center"
                      >
                        Item list is empty
                      </div>
                    </div>
                  </div>
                  <div
                    class="AddRemoveButtons flex gaps"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="footer"
          />
        </div>
      </main>
      <div
        class="HotbarMenu flex column"
      >
        <div
          class="HotbarItems flex column gaps"
        >
          <div
            class="HotbarCell isDraggingOwner animateDown"
            index="0"
          >
            <div
              style="z-index: 12; position: absolute;"
            >
              <div
                class="HotbarIcon contextMenuAvailable"
              >
                <div
                  class="Avatar rounded disabled avatar"
                  id="hotbarIcon-hotbar-icon-catalog-entity"
                  style="width: 40px; height: 40px; background: rgb(5, 1, 130);"
                >
                  Ca
                </div>
              </div>
            </div>
          </div>
          <div
            class="HotbarCell isDraggingOwner animateDown"
            index="1"
          />
          <div
            class="HotbarCell isDraggingOwner animateDown"
            index="2"
          />
          <div
            class="HotbarCell isDraggingOwner animateDown"
            index="3"
          />
          <div
            class="HotbarCell isDraggingOwner animateDown"
            index="4"
          />
          <div
            class="HotbarCell isDraggingOwner animateDown"
            index="5"
          />
          <div
            class="HotbarCell isDraggingOwner animateDown"
            index="6"
          />
          <div
            class="HotbarCell isDraggingOwner animateDown"
            index="7"
          />
          <div
            class="HotbarCell isDraggingOwner animateDown"
            index="8"
          />
          <div
            class="HotbarCell isDraggingOwner animateDown"
            index="9"
          />
          <div
            class="HotbarCell isDraggingOwner animateDown"
            index="10"
          />
          <div
            class="HotbarCell isDraggingOwner animateDown"
            index="11"
          />
        </div>
        <div
          class="HotbarSelector"
        >
          <i
            class="Icon Icon previous material interactive focusable"
            tabindex="0"
          >
            <span
              class="icon"
              data-icon-name="arrow_left"
            >
              arrow_left
            </span>
          </i>
          <div
            class="HotbarIndex"
          >
            <div
              class="badge Badge small clickable"
              id="hotbarIndex"
            >
              1
            </div>
          </div>
          <i
            class="Icon material interactive focusable"
            tabindex="0"
          >
            <span
              class="icon"
              data-icon-name="arrow_right"
            >
              arrow_right
            </span>
          </i>
        </div>
      </div>
      <div
        class="StatusBar"
        data-testid="status-bar"
      >
        <div
          class="leftSide"
          data-testid="status-bar-left"
        />
        <div
          class="rightSide"
          data-testid="status-bar-right"
        />
      </div>
    </div>
    <div
      class="Notifications flex column align-flex-end"
    />
  </div>
</body>
`;
